File: internal/mutationfresh/diff.go
package mutationfresh

import "strings"

// Content-based line diff (spec §11.4.1), a port of templates/mutation-incremental/lib/diff.mjs.
// Both implementations run testdata/mutation-incremental/diff-vectors.json, so a change here needs
// the same change there. Above maxEditDistance lineDiff gives up and the caller treats the file as
// changed everywhere.
const maxEditDistance = 2000

// hunk is one changed region in old and new line numbers (1-based, inclusive). A pure insertion
// has OldEnd == OldStart-1; a pure deletion has NewEnd == NewStart-1.
type hunk struct {
	OldStart int `json:"oldStart"`
	OldEnd   int `json:"oldEnd"`
	NewStart int `json:"newStart"`
	NewEnd   int `json:"newEnd"`
}

func splitLines(text string) []string {
	text = strings.TrimPrefix(text, "\uFEFF")
	text = strings.ReplaceAll(text, "\r\n", "\n")
	text = strings.ReplaceAll(text, "\r", "\n")
	if text == "" {
		return nil
	}
	return strings.Split(strings.TrimSuffix(text, "\n"), "\n")
}

// editOps is Myers' O(ND) shortest edit script with the harness's tie rule: the insertion ("down")
// move when k == -d or v[k-1] < v[k+1], otherwise the deletion ("right") move. trace[d] keeps v for
// diagonals -d-1 .. d+1 as it was before step d.
func editOps(a, b []string) ([]byte, bool) {
	n, m := len(a), len(b)
	off := n + m + 1
	v := make([]int, 2*(n+m)+3)
	var trace [][]int
	for d := 0; d <= maxEditDistance; d++ {
		trace = append(trace, append([]int(nil), v[off-d-1:off+d+2]...))
		for k := -d; k <= d; k += 2 {
			x := v[off+k-1] + 1
			if k == -d || (k != d && v[off+k-1] < v[off+k+1]) {
				x = v[off+k+1]
			}
			y := x - k
			for x < n && y < m && a[x] == b[y] {
				x++
				y++
			}
			v[off+k] = x
			if x >= n && y >= m {
				return backtrack(trace, n, m), true
			}
		}
	}
	return nil, false
}

func backtrack(trace [][]int, n, m int) []byte {
	var ops []byte
	x, y := n, m
	for d := len(trace) - 1; d >= 0; d-- {
		w := trace[d]
		at := func(k int) int { return w[k+d+1] }
		k := x - y
		prevK := k - 1
		if k == -d || (k != d && at(k-1) < at(k+1)) {
			prevK = k + 1
		}
		prevX := at(prevK)
		prevY := prevX - prevK
		for x > prevX && y > prevY {
			ops = append(ops, '=')
			x--
			y--
		}
		if d > 0 {
			if x == prevX {
				ops = append(ops, '+')
			} else {
				ops = append(ops, '-')
			}
		}
		x, y = prevX, prevY
	}
	for i, j := 0, len(ops)-1; i < j; i, j = i+1, j-1 {
		ops[i], ops[j] = ops[j], ops[i]
	}
	return ops
}

// lineDiff returns the changed hunks between two texts after BOM and line-ending normalisation, or
// false when the edit distance exceeds maxEditDistance.
func lineDiff(oldText, newText string) ([]hunk, bool) {
	ops, ok := editOps(splitLines(oldText), splitLines(newText))
	if !ok {
		return nil, false
	}
	var hunks []hunk
	i, j := 0, 0
	open := false
	for _, op := range ops {
		if op == '=' {
			i++
			j++
			open = false
			continue
		}
		if !open {
			hunks = append(hunks, hunk{OldStart: i + 1, OldEnd: i, NewStart: j + 1, NewEnd: j})
			open = true
		}
		cur := &hunks[len(hunks)-1]
		if op == '-' {
			i++
			cur.OldEnd = i
		} else {
			j++
			cur.NewEnd = j
		}
	}
	return hunks, true
}

// hunkIntersects (spec §11.4.3): a mutant [start, end] (old lines) intersects a changed hunk when
// the ranges overlap, and a pure insertion when it contains the insertion point strictly inside.
func hunkIntersects(start, end int, h hunk) bool {
	if h.OldEnd >= h.OldStart {
		return start <= h.OldEnd && end >= h.OldStart
	}
	after := h.OldStart - 1
	return start <= after && end >= after+1
}

File: templates/mutation-incremental/lib/diff.mjs
// Content-based line diff (spec §11.4.1), shared with the Go gate. Both implementations run
// testdata/mutation-incremental/diff-vectors.json, so a change here needs the same change there.
// Above MAX_EDIT_DISTANCE lineDiff returns null and the planner forces the file whole.
export const MAX_EDIT_DISTANCE = 2000;

export function normalizeText(text) {
  const noBom = text.startsWith('\ufeff') ? text.slice(1) : text;
  return noBom.replace(/\r\n?/g, '\n');
}

export function splitLines(text) {
  const t = normalizeText(text);
  if (t === '') return [];
  const lines = t.split('\n');
  if (lines[lines.length - 1] === '') lines.pop();
  return lines;
}

// Myers' O(ND) shortest edit script. Ties take the insertion ("down") move when k === -d or
// v[k-1] < v[k+1], otherwise the deletion ("right") move, so the result is deterministic.
// trace[d] keeps v for diagonals -d-1 .. d+1 as it was before step d.
function editOps(a, b) {
  const n = a.length;
  const m = b.length;
  const off = n + m + 1;
  const v = new Int32Array(2 * (n + m) + 3);
  const trace = [];
  for (let d = 0; d <= MAX_EDIT_DISTANCE; d++) {
    trace.push(v.slice(off - d - 1, off + d + 2));
    for (let k = -d; k <= d; k += 2) {
      let x = k === -d || (k !== d && v[off + k - 1] < v[off + k + 1]) ? v[off + k + 1] : v[off + k - 1] + 1;
      let y = x - k;
      while (x < n && y < m && a[x] === b[y]) {
        x++;
        y++;
      }
      v[off + k] = x;
      if (x >= n && y >= m) return backtrack(trace, n, m);
    }
  }
  return null;
}

function backtrack(trace, n, m) {
  const ops = [];
  let x = n;
  let y = m;
  for (let d = trace.length - 1; d >= 0; d--) {
    const w = trace[d];
    const at = (k) => w[k + d + 1];
    const k = x - y;
    const prevK = k === -d || (k !== d && at(k - 1) < at(k + 1)) ? k + 1 : k - 1;
    const prevX = at(prevK);
    const prevY = prevX - prevK;
    while (x > prevX && y > prevY) {
      ops.push('=');
      x--;
      y--;
    }
    if (d > 0) ops.push(x === prevX ? '+' : '-');
    x = prevX;
    y = prevY;
  }
  return ops.reverse();
}

export function lineDiff(oldText, newText) {
  const a = splitLines(oldText);
  const b = splitLines(newText);
  const ops = editOps(a, b);
  if (ops === null) return null;
  const hunks = [];
  const oldToNew = [null];
  let i = 0;
  let j = 0;
  let cur = null;
  for (const op of ops) {
    if (op === '=') {
      i++;
      j++;
      oldToNew[i] = j;
      cur = null;
      continue;
    }
    if (cur === null) {
      cur = { oldStart: i + 1, oldEnd: i, newStart: j + 1, newEnd: j };
      hunks.push(cur);
    }
    if (op === '-') {
      i++;
      cur.oldEnd = i;
      oldToNew[i] = null;
    } else {
      j++;
      cur.newEnd = j;
    }
  }
  return { hunks, oldToNew };
}

// Spec §11.4.3: a mutant [start, end] (old lines) intersects a changed hunk when the ranges overlap,
// and a pure insertion when it contains the insertion point (strictly inside the mutant's span).
export function hunkIntersects(start, end, h) {
  if (h.oldEnd >= h.oldStart) return start <= h.oldEnd && end >= h.oldStart;
  const after = h.oldStart - 1;
  return start <= after && end >= after + 1;
}
